state/modelconfig.go

```diff
@@ -201,37 +209,42 @@ func (st *State) UpdateModelConfigDefaultValues(updateAttrs map[string]interface
 	// applied as a delta to what's on disk; if there has
 	// been a concurrent update, the change may not be what
 	// the user asked for.
-
-	// Attempt to validate against the current old model and the new model, that
-	// should be enough to verify the config against.
-	// If there are additional fields in the config, then this should be fine
-	// and should not throw a validation error.
-	model, err := st.Model()
-	if err != nil {
-		return errors.Trace(err)
+	updateAttrs = config.CoerceForStorage(updateAttrs)
+	settings.Update(updateAttrs)
+	for _, r := range removeAttrs {
+		settings.Delete(r)
 	}
-	oldConfig, err := model.ModelConfig()
-	if err != nil {
+
+	if err := st.validateModelDefaults(settings.Map()); err != nil {
 		return errors.Trace(err)
 	}
-	validCfg, err := st.buildAndValidateModelConfig(updateAttrs, removeAttrs, oldConfig)
+
+	_, err = settings.Write()
+	return err
+}
+
+// validateModelDefaults does what provider-agnostic validation we can on the
+// input configuration attributes, which are candidates for setting as model
+// defaults.
+// The attributes are not mutated, but we copy them and set some values
+// required for validation to succeed. This is due to validation assuming a
+// fully formed model config, rather than a subset of defaults.
+func (st *State) validateModelDefaults(attrs map[string]any) error {
+	forValidation := copyMap(attrs, nil)
+
+	forValidation["name"] = "valid-model-name"
+	forValidation["type"] = model.IAAS
+	forValidation["uuid"] = st.ModelUUID()
+
+	cfg, err := config.New(config.NoDefaults, forValidation)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	validAttrs := validCfg.AllAttrs()
-	for k := range updateAttrs {
-		if v, ok := validAttrs[k]; ok {
-			updateAttrs[k] = v
-		}
-	}
 
-	updateAttrs = config.CoerceForStorage(updateAttrs)
-	settings.Update(updateAttrs)
-	for _, r := range removeAttrs {
-		settings.Delete(r)
+	if err := config.Validate(cfg, &config.Config{}); err != nil {
```

Comment:
config Validate() accepts old config being passed in as nil and we even have existing calls sites using that